dask/dataframe/core.py

```diff
@@ -1029,13 +1029,17 @@ def elemwise(op, *args, **kwargs):
     else:
         op2 = op
 
-    if not all(df.divisions == dfs[0].divisions for df in dfs):
-        msg = 'All dask.Dataframe and dask.Series must have same divisions'
+    if not all(df.divisions[0] == dfs[0].divisions[0] for df in dfs):
+        msg = 'All divisions must have the same left edge'
         raise ValueError(msg)
-    if not all(df.npartitions == dfs[0].npartitions for df in dfs):
-        msg = 'All dask.Dataframe and dask.Series must have same npartitions'
+    if not all(df.divisions[-1] == dfs[0].divisions[-1] for df in dfs):
+        msg = 'All divisions must have the same right edge'
```

Comment:
Following is a small example.

```
import numpy as np
import pandas as pd
import dask.dataframe as dd

df1 = pd.DataFrame(np.arange(5))
ddf1 = dd.from_pandas(df1, 2)
df2 = pd.DataFrame(np.arange(5), index=[3, 4, 5, 6, 7])
ddf2 = dd.from_pandas(df2, 2)

df1 + df2
#     0
# 0 NaN
# 1 NaN
# 2 NaN
# 3   3
# 4   5
# 5 NaN
# 6 NaN
# 7 NaN

(ddf1 + ddf2).compute()
#     0
# 0 NaN
# 1 NaN
# 2 NaN
# 3 NaN
# 3 NaN
# 4   5
# 5 NaN
```

I **assume** align partition should handle these cases, but not yet checking in detail.

```
dfs, _, _ = dd.multi.align_partitions(ddf1, ddf2)
dfs
# [dd.DataFrame<repartition-merge-f214ab44a05c6216b25fdf8a6bc8d072, divisions=(0, 3, 4)>, 
#  dd.DataFrame<repartition-merge-e252f94976b0c47b7523172e030c1a32, divisions=(3, 4, 6, 7)>]
```